Improve: allow `isActive(...)` and `exists(...)` to use an ID number

SlySven opened the pull request:
#### Brief overview of PR changes/additions
With this PR passing an actual number as the first argument will use it as an ID number (provided it rounds to a zero or greater integer) and return a `1` or `0` if the corresponding Mudlet item is "active" or not / exists or not (though like the unmodified one that returns a count of items of the selected type that are active it does not consider whether all the item's "ancestors" are active as well before including an item in the count).

This can be used to identify the active status of temporary items (which are only identified by an ID number) - provided of course that they are around long enough to be checked. If the item does NOT exist then a `nil` and an error message will be returned instead.

N. B. `isActive(...)` will return `0` **if the item (when given an ID number) or items (when given a name) does NOT exist** - this might not be what is expected but is how the previous "name-only" form worked. OTOH ID numbers less than zero are never (:crossed_fingers:) going to occur and those will provoke a `nil` + error message if used.

#### Motivation for adding to Mudlet
This came about as a result of a comment in the Discord \#help channel:
> Kalrykh — 2022/10/29 at 15:15 UTC:
>
> I've spent the past few(maybe more than a few) hours setting up some things with tempTriggers and tempTimers with regular triggers.  Is there a way to show all active temp stuff? i can clear everything out by cycling mudlet, but I'd like to see if I've been running around generating tempTriggers/Timers without realizing it due to sloppy code on my part.  if i don't know, they'll just keep getting generated by the regular triggers i'm using even after cycling mudlet.

> SlySven — 2022/10/29 at 1526 UTC:
> Humm, it seems that `isActive(itemName, itemType)` only works with a name (string) and not an ID number - so there is no way to find out the state of a temporary item.

#### Other info (issues closed, discussion etc)
This ought to help with the sort of issue that the OP raised (#6397) though a script using this revised function would, I think, be needed to do what is wanted there.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Improve: allow isActive(...) to use an ID number

With this PR passing an actual number as the first argument will use it as
an ID number (provided it rounds to a zero or greater integer) and return
a `1` or `0` if the corresponding Mudlet item is "active" or not (though
like the unmodified one that returns a count of items of the selected type
that are active it does not consider whether all the item's "ancestors" are
active as well before including an item in the count).

This can be used to identify the active status of temporary items (which
are only identified by an ID number) - provided of course that they are
around long enough to be checked. If the item does NOT exist then a `nil`
and an error message will be returned instead.

This ought to help with the sort of issue that #6397 raises though a script
using this revised function would, I think, be needed to do what is wanted
there.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

Kebap commented:
Not sure if this addition is worthwhile really.
Usually players put their temp items in tables to know how many there are.
Maybe I am missing something.

Then again, would rather see a new function than overload existing API.
It is more newbie friendly and easier to code and easier to document.
I see hardly any advantage of using the same function name again.

SlySven commented:
**OTOH It is *impossible* to determine the *state* of an existing temporary item.** This is independent of whether the end-user generated it themselves and remembered/recorded the ID or not. Also, reusing/overloading the function name reduces the number of *different* functions that we have to document but extends the functionality in an, I *think*, logical/useful manner.

Kebap commented:
>  It is impossible to determine the state of an existing temporary item

Fair enough.

> reusing/overloading the function name reduces the number of different functions that we have

My point exactly. I would prefer to have many small concise functions which will do one thing exactly and well, rather than very few very complex functions which can do a dozen different things, if you know how to wield them for that case.

While overloading is possible in c++ (but not c itself) it is rather uncommon in Lua, and I have pointed out some problems it may cause down the line already.

SlySven commented:
Closing in favour of #6403...

SlySven commented:
Reopened as this approach was deemed more acceptable in this case than https://github.com/Mudlet/Mudlet/pull/6403...

SlySven committed:
Merge branch 'development' into Improve_allow_isActive_toUseIdNumber

SlySven commented:
Need to revise the `exists(...)` function to also take a numeric first argument as well.

Kebap commented:
Can't review this code is above my level but some of my comments from the other proposal may hold true still.

SlySven committed:
Merge branch 'development' into Improve_allow_isActive_toUseIdNumber

SlySven committed:
Revise: also allow exists(...) to take a numeric ID

This has been selected to be the more desirable way of doing things
compared to the alternative of #6403.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Merge branch 'development' into Improve_allow_isActive_toUseIdNumber

vadi2 commented on src/TLuaInterpreter.cpp:
```diff
@@ -219,7 +219,11 @@ TLuaInterpreter::~TLuaInterpreter()
 //    bool showOnTop = getVerifiedBool(L, "createMapLabel", 14, "showOnTop", true);
 //
-// The "isOptional" parameter is optional, and will default to not-optional parameters! :)
+// The "isOptional" parameter is optional but modifies the error message to say
+// that an argument is optional and it will default to not-optional parameters!
+// HOWEVER it does not actually handle the absence of an argument that is
+// supposed to BE optional - that has to be done by the caller before it
+// makes the call... 8-P
```
This is a bit of a bummer but I think the logic here might confuse others, especially not native English speakers

SlySven replied:
These are *internal* use functions though - only seen by us developers - who will need to remember this detail! :imp:

vadi2 replied:
We have non-native English speaking developers (and would love more), can you reword this to be simpler?

SlySven replied:
@vadi2
```diff
-// The "isOptional" parameter is optional but modifies the error message to say
-// that an argument is optional and it will default to not-optional parameters!
-// HOWEVER it does not actually handle the absence of an argument that is
-// supposed to BE optional - that has to be done by the caller before it
-// makes the call... 8-P
+// The (optional here) "isOptional" parameter only modifies the error message
+// given to the user to say that the parameter need not be provided. It does not
+// (until someone works out how to!) allow this function to deal with the argument
+// at the given pos not being present - so it is up to the function that calls this to
+// check for that BEFORE trying to call this function. This applies to all the
+// "getVerifiedXxxx" functions.
```

vadi2 replied:
Sounds good!

vadi2 commented on src/TLuaInterpreter.cpp:
```diff
@@ -233,6 +237,26 @@ bool TLuaInterpreter::getVerifiedBool(lua_State* L, const char* functionName, co
+    if (lua_type(L, pos) == LUA_TNUMBER) {
+        // use lua_tonumber(...) and round because lua_tointeger(...) can return
+        // oversized values (long long int?) on Windows which do not always fit
+        // into an int:
```
👍

SlySven committed:
Merge branch 'development' into Improve_allow_isActive_toUseIdNumber

SlySven commented:
:books: Documentation in Main wiki updated to (correctly) refer to handling ID numbers - the prior text incorrectly suggested that ID numbers already worked. See **[exists(...)](https://wiki.mudlet.org/w/Manual:Mudlet_Object_Functions#exists)** and **[isActive(...)](https://wiki.mudlet.org/w/Manual:Mudlet_Object_Functions#isActive)**.

~~Will require a further edit to replace **Mudlet tbc** with **Mudlet 4.17.0** as I am not assuming anything at this time...!~~ - *Done.*
